packages/react-dom-bindings/src/events/plugins/ScrollEndEventPlugin.js

```diff
@@ -0,0 +1,212 @@
+      'onScrollEnd',
+      'scrollend',
+      null,
+      nativeEvent, // This will be the "scroll" event.
+      target,
+    );
+    dispatchQueue.push({event, listeners});
+  }
+  batchedUpdates(runEventInBatch, dispatchQueue);
+}
+
+function runEventInBatch(dispatchQueue: DispatchQueue) {
+  processDispatchQueue(dispatchQueue, 0);
+}
+
+function fireScrollEnd(
+  targetInst: Fiber,
+  nativeEvent: AnyNativeEvent,
+  nativeEventTarget: EventTarget,
+): void {
+  clearScrollEndTimer(nativeEventTarget);
+  if (isMouseDown || isTouchStarted) {
+    // If mouse or touch is down, try again later in case this is due to having an
+    // active scroll but it's not currently moving.
+    debounceScrollEnd(targetInst, nativeEvent, nativeEventTarget);
+    return;
+  }
+  manualDispatchScrollEndEvent(targetInst, nativeEvent, nativeEventTarget);
+}
+
+// When scrolling slows down the frequency of new scroll events can be quite low.
+// This timeout seems high enough to cover those cases but short enough to not
+// fire the event way too late.
+const DEBOUNCE_TIMEOUT = 200;
+
+function debounceScrollEnd(
+  targetInst: null | Fiber,
+  nativeEvent: AnyNativeEvent,
+  nativeEventTarget: EventTarget,
+) {
+  const existingTimer = getScrollEndTimer(nativeEventTarget);
+  if (existingTimer != null) {
+    clearTimeout(existingTimer);
+  }
+  if (targetInst !== null) {
+    const newTimer = setTimeout(
+      fireScrollEnd.bind(null, targetInst, nativeEvent, nativeEventTarget),
+      DEBOUNCE_TIMEOUT,
+    );
+    setScrollEndTimer(nativeEventTarget, newTimer);
+  }
+}
+
+/**
+ * This plugin creates an `onScrollEnd` event polyfill when the native one
+ * is not available.
+ */
+function extractEvents(
+  dispatchQueue: DispatchQueue,
+  domEventName: DOMEventName,
+  targetInst: null | Fiber,
+  nativeEvent: AnyNativeEvent,
+  nativeEventTarget: null | EventTarget,
+  eventSystemFlags: EventSystemFlags,
+  targetContainer: null | EventTarget,
+) {
+  if (!enableScrollEndPolyfill) {
+    return;
+  }
+
+  const inCapturePhase = (eventSystemFlags & IS_CAPTURE_PHASE) !== 0;
+
+  if (domEventName !== 'scrollend') {
+    if (!isScrollEndEventSupported && inCapturePhase) {
+      switch (domEventName) {
+        case 'scroll': {
+          if (nativeEventTarget !== null) {
+            debounceScrollEnd(targetInst, nativeEvent, nativeEventTarget);
+          }
+          break;
+        }
+        case 'touchstart': {
+          isTouchStarted = true;
+          break;
+        }
+        case 'touchcancel':
+        case 'touchend': {
+          // Note we cannot use pointer events for this because they get
+          // cancelled when native scrolling takes control.
+          isTouchStarted = false;
+          break;
+        }
+        case 'mousedown': {
+          isMouseDown = true;
+          break;
+        }
+        case 'mouseup': {
+          isMouseDown = false;
+          break;
+        }
```

Comment:
Also shouldn't assume all mouse events are for left clicks, and `mouseup` won't fire after a drag starts.

```suggestion
        case 'mousedown': {
          if (((nativeEvent: any): MouseEvent).button === 0) {
            isMouseDown = true;
          }
          break;
        }
        case 'mouseup':
        case 'dragend': {
          if (((nativeEvent: any): MouseEvent).button === 0) {
            isMouseDown = false;
          }
          break;
        }
```